ffi/dart/tool/build_libs.dart

```diff
@@ -2,120 +2,251 @@
 // called LICENSE at the top level of the ICU4X source tree
 // (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
 
-import 'package:native_assets_cli/native_assets_cli.dart';
 import 'dart:io';
 
+import 'package:args/args.dart';
+import 'package:native_assets_cli/code_assets.dart';
+import 'package:path/path.dart' as path;
+
+const crateName = 'icu_capi';
+
 Future<void> main(List<String> args) async {
-  if (args.isEmpty) {
-    print(
-        'Usage: ${Platform.script.pathSegments.last} <out_dir> <target:${Target.values}> <link mode: ${LinkMode.values}> <cargo features>');
+  final defaultFeatures = ['default_components', 'compiled_data'];
+  var fileKey = 'file';
+  var osKey = 'os';
+  var architectureKey = 'architecture';
+  var simulatorKey = 'simulator';
+  var compileTypeKey = 'compile_type';
+  var cargoFeaturesKey = 'cargo_features';
+  final argParser =
+      ArgParser()
+        ..addOption(fileKey, mandatory: true)
+        ..addOption(
+          compileTypeKey,
+          allowed: ['static', 'dynamic'],
+          mandatory: true,
+        )
+        ..addFlag(simulatorKey, defaultsTo: false)
+        ..addOption(osKey, mandatory: true)
+        ..addOption(architectureKey, mandatory: true)
+        ..addMultiOption(cargoFeaturesKey, defaultsTo: defaultFeatures);
+
+  ArgResults parsed;
+  try {
+    parsed = argParser.parse(args);
+  } catch (e) {
+    print('Error parsing $args');
+    print(argParser.usage);
     exit(1);
   }
-  final out = Uri.file(args[0]).toFilePath();
-  final target = Target.values.firstWhere((t) => t.toString() == args[1]);
-  final linkMode = LinkMode.values.firstWhere((l) => l.toString() == args[2]);
+  final buildStatic = parsed.option(compileTypeKey)! == 'static';
+  final simulator = parsed.flag(simulatorKey);
+  final targetOS = OS.values.firstWhere((o) => o.name == parsed.option(osKey)!);
+  final targetArchitecture = Architecture.values.firstWhere(
+    (o) => o.name == parsed.option(architectureKey)!,
+  );
 
-  final defaultFeatures = ['default_components', 'compiled_data'];
-  final cargoFeatures = args.elementAtOrNull(3)?.split(',') ?? defaultFeatures;
+  final out = Uri.file(parsed.option(fileKey)!);
 
-  await buildLib(target, linkMode, cargoFeatures, out);
-}
-
-Future<void> buildLib(
-  Target target,
-  LinkMode linkMode,
-  Iterable<String> cargoFeatures,
-  String outName,
-) async {
-  var root = Platform.script.toFilePath().split('ffi')[0];
-  root = root.substring(0, root.length - 1); // trim trailing slash
+  final cargoFeatures = parsed.multiOption(cargoFeaturesKey);
 
-  print('Building $outName');
+  // We assume that the first folder to contain a cargo.toml above the
+  // output directory is the directory containing the ICU4X code.
+  Directory icu4xPath = recurseToParentRustCrate(File.fromUri(Platform.script));
```

Comment:
you can further inline this. the result from this is always passed down to `buildLibraryFromInput` and not used anywhere else, so that function can just detect the working directory internally

Reply:
It's passed to `buildLib`, which has two calling sites